File: packages/bdd/src/ast.ts
import type { Span } from './span.js'

export type Heading = {
  readonly kind: 'heading'
  readonly level: 1 | 2 | 3 | 4 | 5 | 6
  readonly text: string
  readonly span: Span
}

export type Paragraph = {
  readonly kind: 'paragraph'
  readonly text: string
  readonly span: Span
  readonly inlineMap: ReadonlyArray<InlineOffset>
}

export type InlineOffset = {
  readonly textOffset: number
  readonly sourceOffset: number
}

export type ListItem = {
  readonly kind: 'list_item'
  readonly text: string
  readonly span: Span
  readonly inlineMap: ReadonlyArray<InlineOffset>
  readonly ordered: boolean
  readonly markerSpan: Span
}

export type Blockquote = {
  readonly kind: 'blockquote'
  readonly text: string
  readonly span: Span
  readonly inlineMap: ReadonlyArray<InlineOffset>
}

export type Row = { readonly cells: ReadonlyArray<string>; readonly span: Span }

export type Table = {
  readonly kind: 'table'
  readonly span: Span
  readonly header: Row
  readonly rows: ReadonlyArray<Row>
}

export type Fence = {
  readonly kind: 'fence'
  readonly span: Span
  readonly info: string
  readonly body: string
  readonly bodySpan: Span
}

export type ThematicBreak = {
  readonly kind: 'thematic_break'
  readonly span: Span
}

export type Block = Heading | Paragraph | ListItem | Blockquote | Table | Fence | ThematicBreak

export type Example = {
  readonly name: string
  readonly span: Span
  readonly headingSpan: Span
  readonly body: ReadonlyArray<Block>
}

export type Bdd = {
  readonly path: string
  readonly source: string
  readonly examples: ReadonlyArray<Example>
  readonly orphanAttachments: ReadonlyArray<Table | Fence>
}

File: python/packages/var-core/src/var_core/ast.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Union

from var_core.span import Span


@dataclass(frozen=True, slots=True)
class InlineOffset:
    text_offset: int
    source_offset: int


@dataclass(frozen=True, slots=True)
class Heading:
    kind: str = "heading"
    level: int = 0
    text: str = ""
    span: Span | None = None

    def __post_init__(self) -> None:
        if self.span is None:
            raise TypeError("span is required")


@dataclass(frozen=True, slots=True)
class Paragraph:
    kind: str = "paragraph"
    text: str = ""
    span: Span | None = None
    inline_map: tuple[InlineOffset, ...] = ()

    def __post_init__(self) -> None:
        if self.span is None:
            raise TypeError("span is required")


@dataclass(frozen=True, slots=True)
class ListItem:
    kind: str = "list_item"
    text: str = ""
    span: Span | None = None
    inline_map: tuple[InlineOffset, ...] = ()
    ordered: bool = False
    marker_span: Span | None = None

    def __post_init__(self) -> None:
        if self.span is None:
            raise TypeError("span is required")
        if self.marker_span is None:
            raise TypeError("marker_span is required")


@dataclass(frozen=True, slots=True)
class Blockquote:
    kind: str = "blockquote"
    text: str = ""
    span: Span | None = None
    inline_map: tuple[InlineOffset, ...] = ()

    def __post_init__(self) -> None:
        if self.span is None:
            raise TypeError("span is required")


@dataclass(frozen=True, slots=True)
class Row:
    cells: tuple[str, ...] = ()
    cell_spans: tuple[Span, ...] = ()
    span: Span | None = None

    def __post_init__(self) -> None:
        if self.span is None:
            raise TypeError("span is required")


@dataclass(frozen=True, slots=True)
class Table:
    kind: str = "table"
    span: Span | None = None
    header: Row | None = None
    rows: tuple[Row, ...] = ()

    def __post_init__(self) -> None:
        if self.span is None:
            raise TypeError("span is required")
        if self.header is None:
            raise TypeError("header is required")


@dataclass(frozen=True, slots=True)
class Fence:
    kind: str = "fence"
    span: Span | None = None
    info: str = ""
    body: str = ""
    body_span: Span | None = None

    def __post_init__(self) -> None:
        if self.span is None:
            raise TypeError("span is required")
        if self.body_span is None:
            raise TypeError("body_span is required")


@dataclass(frozen=True, slots=True)
class ThematicBreak:
    kind: str = "thematic_break"
    span: Span | None = None

    def __post_init__(self) -> None:
        if self.span is None:
            raise TypeError("span is required")


Block = Union[Heading, Paragraph, ListItem, Blockquote, Table, Fence, ThematicBreak]


@dataclass(frozen=True, slots=True)
class Example:
    scope_stack: tuple[str, ...] = ()
    span: Span | None = None
    body: tuple[Block, ...] = ()

    def __post_init__(self) -> None:
        if self.span is None:
            raise TypeError("span is required")


@dataclass(frozen=True, slots=True)
class VarDoc:
    path: str = ""
    source: str = ""
    examples: tuple[Example, ...] = ()
    orphan_attachments: tuple[Union[Table, Fence], ...] = ()
